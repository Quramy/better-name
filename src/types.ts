export type $DiffKey<T, U> = T extends U ? never : T;
export type $Diff<T, U> = Pick<T, $DiffKey<keyof T, keyof U>>;
export type $Optional<T> = { [P in keyof T]?: T[P] };
export type $PartialOptional<T, U> = $Diff<T, U> & $Optional<U>

export type FileId = string;

export interface FileRef {
  readonly id: FileId;
  readonly path: string;
}

export interface SourceReader {
  read(file: FileRef): Promise<string>;
}

export interface SourceWriter {
  write(file: FileRef, source: string): Promise<void>;
}

export interface SourceRemover {
  delete(file: FileRef): Promise<void>;
}

export interface DocumentRef {
  getFile(): FileRef;
  getDoc(): DocumentEntity;
  detach(): void;
  move(newFile: FileRef): Promise<this>;
}

export interface TransformOptions {
  from: string;
  to: string;
}

export interface DocumentEntityCreateOptions {
  fileRef: FileRef,
  fileMappingOptions?: FileMappingOptions,
}

export interface DocumentEntity {
  readonly fileRef: FileRef;
  readonly isDirty: boolean;
  reader: SourceReader;
  writer: SourceWriter;
  parse(): Promise<this>;
  transformPreceding(to: string): this;
  transformFollowing(opt: TransformOptions): this;
  flush(force?: boolean): Promise<this>;
  move(newFile: FileRef): Promise<this>;
}

export type RootImportConfig = {
  rootPathSuffix?: string;
  rootPathPrefix?: string;
};

export type FileMappingOptions = {
  rootImport?: RootImportConfig[];
}

export type FindQuery = {
  start: string;
};

export interface Project {
  getProjectDir(): string;
  getFileMappingOptions(): FileMappingOptions;
  getDocumentsList(): Promise<DocumentRef[]>;
  findOne(fileId: FileId): Promise<{ found: DocumentRef | undefined, rest: DocumentRef[] }>;
  find(query: FindQuery): Promise<{ found: DocumentRef[], rest: DocumentRef[] }>;
}
