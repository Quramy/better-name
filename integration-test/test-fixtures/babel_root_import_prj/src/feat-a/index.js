import { awesome } from '~/common/util';

awesome();
