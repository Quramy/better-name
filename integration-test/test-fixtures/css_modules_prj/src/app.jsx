import React from 'react';
import { render } from 'react-dom';
import Hoge from './components/Hoge/Hoge';

render(<Hoge />, document.getElementById('app'));
