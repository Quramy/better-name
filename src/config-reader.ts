import * as fs from "fs";
import * as path from "path";

import {
  FileMappingOptions,
  RootImportConfig,
} from "./types";

import {
  exists,
  readFileAsJson,
} from "./file-util";

export type BabelrcType = {
  plugins?: (
    {
      0: string,
      1: any,
    } | {
      0: "babel-root-imoprt",
      1: RootImportConfig[],
    } | string
  )[];
};

export function extractRootImportConfigFromBabelrc(babelrc: BabelrcType): RootImportConfig[] {
  if (!babelrc.plugins) return [];
  let conf: RootImportConfig[] = [];
  babelrc.plugins.forEach(p => {
    if (typeof p === "string") return;
    if (p[0] !== "babel-root-import") return;
    conf = p[1];
  })
  return conf;
}

export async function readRootImportConfig(rootDir: string) {
  try {
    let ret: RootImportConfig[] = [];
    if (exists(rootDir, "package.json")) {
      const pkg = await readFileAsJson(rootDir, "package.json");
      if (pkg["rootImport"]) {
        return pkg["rootImport"] as RootImportConfig[];
      }
    }
    if (exists(rootDir, ".babelrc")) {
      const babelrc = await readFileAsJson(rootDir, ".babelrc");
      return extractRootImportConfigFromBabelrc(babelrc);
    } 
    return [] as RootImportConfig[];
  } catch(err) {
    return [] as RootImportConfig[];
  }
}
