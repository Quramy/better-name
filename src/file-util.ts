import * as path from "path";
import * as fs from "fs";
import { FileRef, SourceReader, SourceWriter } from "./project";

function toFileId(filePath: string) {
  return filePath.replace(/\\/g, "/");
}

export class DefaultFileRef implements FileRef {

  readonly id: string;
  readonly path: string;

  constructor(filePath: string, root?: string) {
    this.path = root ? path.resolve(root, filePath) : filePath;
    this.id = toFileId(root && path.isAbsolute(filePath) ? path.relative(root, filePath) : filePath);
  }
}

export class FileSourceReader implements SourceReader {
  read(file: FileRef): Promise<string> {
    return new Promise((res, rej) => {
      fs.readFile(file.path, "utf-8", (err, content) => {
        if (err) return rej(err);
        res(content);
      });
    });
  }
}

export class FileSourceWriter implements SourceWriter {
  write(file: FileRef, source: string): Promise<void> {
    return new Promise((res, rej) => {
      fs.writeFile(file.path, source, "utf-8", (err) => {
        if (err) return rej(err);
        res();
      });
    });
  }
}
