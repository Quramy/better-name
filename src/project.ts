import * as glob from "glob";
import * as path from "path";
import * as fs from "fs";

import {
  $PartialOptional,
  Project,
  DocumentRef,
  DocumentEntity,
  FileRef,
  SourceReader,
  SourceWriter,
  SourceRemover,
  FileMappingOptions,
} from "./types";

import { DefaultFileRef, FileSourceReader, FileSourceWriter, RimrafAdapter } from "./file-util";
import { shouldBeReplaced, shouldBeReplacedWithModuleMove } from "./functions";
import { BabylonDocmentEntity } from "./docEntity";

import {
  readRootImportConfig,
} from "./config-reader";

export class DefaultProject implements Project {
  private _docRefList?: DocumentRef[];
  protected reader: SourceReader = new FileSourceReader();
  protected writer: SourceWriter = new FileSourceWriter();
  protected remover: SourceRemover = new RimrafAdapter();

  constructor(
    private _config: AllProjectOptions,
  ) {
  }

  getProjectDir() {
    return this._config.rootDir;
  }

  getFileMappingOptions() {
    return this._config.fileMapping;
  }

  getDocumentsList() {
    if (this._docRefList) {
      return Promise.resolve(this._docRefList);
    }
    return new Promise<DocumentRef[]>((resolve, reject) => {
      glob(this._config.pattern, { cwd: this._config.rootDir }, (err, files) => {
        if (err) return reject(err);
        resolve(files.map(f => {
          const fileRef = new DefaultFileRef(f, this._config.rootDir);
          return new DefaultDocumentRef({
            fileRef,
            reader: this.reader,
            writer: this.writer,
            remover: this.remover,
            fileMappingOptions: this._config.fileMapping,
          });
        }));
      });
    });
  }
}

export type AllProjectOptions = {
  rootDir: string;
  pattern: string;
  fileMapping: FileMappingOptions;
};

export const defaultProjectConfig = {
  pattern: "src/**/*.js",
  fileMapping: { },
}

export type ProjectOptions = $PartialOptional<AllProjectOptions, typeof defaultProjectConfig>;

export async function createProject<X extends DefaultProject>(k: typeof DefaultProject, configuration: ProjectOptions) {
  const { rootDir } = configuration;
  const rootImport = await readRootImportConfig(rootDir);
  const conf = { ...defaultProjectConfig, ...configuration }
  conf.fileMapping = {
    rootImport: rootImport,
  };
  const prj = new k({ 
    ...conf,
  }) as X;
  return prj;
}

export function createDefaultProject(configuration: ProjectOptions) {
  return createProject<DefaultProject>(DefaultProject, configuration);
}

export type DefaultDocumentRefCreateOptioons = {
  fileRef: FileRef,
  reader: SourceReader,
  writer: SourceWriter,
  remover: SourceRemover,
  fileMappingOptions: FileMappingOptions,
};

export class DefaultDocumentRef implements DocumentRef {

  constructor(private _opt: DefaultDocumentRefCreateOptioons) {
  }

  private _ref?: DocumentEntity | null;

  getRef() {
    if (this._ref) return this._ref;
    const {
      reader,
      writer,
      remover,
      ...rest
    } = this._opt;
    const ref = new BabylonDocmentEntity({ ...rest });
    ref.reader = reader;
    ref.writer = writer;
    ref.remover = remover;
    this._ref = ref;
    return this._ref;
  }

  detach(): void {
    this._ref = null;
  }

}
