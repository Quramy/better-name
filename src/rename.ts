import { Project } from "./types";
import { DefaultFileRef } from "./file-util";
import * as path from "path";

function createToPath(from: string, to: string) {
  const f = path.basename(from);
  if (path.extname(to) === "") {
    return path.join(to, f);
  } else {
    return to;
  }
}

export async function rename(prj: Project, fromPath: string, toPath: string) {
  const fromPrefix = new DefaultFileRef(fromPath, prj.getProjectDir());
  const { found } = await prj.find({ start: fromPrefix.id });
  if (!found.length) return prj;
  await found.reduce(async (prev, file) => {
    await prev;
    const { found, rest } = await prj.findOne(file.getFile().id);
    const to = new DefaultFileRef(createToPath(file.getFile().path, toPath), prj.getProjectDir());
    if (found) {
      await found.getDoc().parse();
      found.getDoc().transformPreceding(to.id);
    }
    await Promise.all(rest.map(async docRef => {
      const doc = docRef.getDoc();
      await doc.parse();
      doc.transformFollowing({ from: file.getFile().id, to: to.id });
      await doc.flush();
      return;
    }));
    if (!found) return prj;
    await found.move(to);
    return prj;
  }, Promise.resolve() as Promise<any>);
  return prj;
}
