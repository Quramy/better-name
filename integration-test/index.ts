import { FileRef, SourceWriter, SourceRemover, } from "../src/types";
import { DefaultProject, createProject, AllProjectOptions } from "../src/project";
import { FileSourceReader } from "../src/file-util";
import { rename } from "../src/rename";
import * as path from "path";

class FixtureIo extends FileSourceReader implements SourceWriter, SourceRemover {
  contentsMap = new Map<string, string>();

  write(file: FileRef, source: string): Promise<void> {
    this.contentsMap.set(file.id, source);
    return Promise.resolve();
  }

  read(file: FileRef) {
    if (this.contentsMap.has(file.id)) {
      return Promise.resolve(this.contentsMap.get(file.id) as string);
    } else {
      return super.read(file);
    }
  }

  delete(file: FileRef) {
    this.contentsMap.delete(file.id);
    return Promise.resolve();
  }

  dump() {
    const keys = Array.from(this.contentsMap.keys()).sort();
    return keys.map(k => {
      return `
// id: ${k}

${this.contentsMap.get(k)}`;
    }).join("\n");
  }
}

class TestProject extends DefaultProject {
  io: FixtureIo;

  constructor(args: any) {
    super(args);
    const fixtureIo = new FixtureIo();
    this.io = fixtureIo;
    this.reader = fixtureIo;
    this.writer = fixtureIo;
    this.remover = fixtureIo;
  }

  getSnapshot() {
    return this.io.dump();
  }
}

describe("integration test", () => {

  it("simple_babel_prj", async done => {
    const rootDir = path.join(__dirname, "test-fixtures/simple_babel_prj");
    const prj = await createProject<TestProject>(TestProject, { rootDir });
    try {
      await rename(prj, path.join(rootDir, "src/core/target.js"), path.join(rootDir, "src/feat/dest.js"));
      expect(prj.getSnapshot()).toMatchSnapshot();
      done();
    } catch (err) {
      done(err);
    }
  });

  it("simple_ts_prj", async done => {
    const rootDir = path.join(__dirname, "test-fixtures/simple_ts_prj");
    const prj = await createProject<TestProject>(TestProject, { rootDir });
    try {
      await rename(prj, path.join(rootDir, "src/core/target.ts"), path.join(rootDir, "src/feat/dest.ts"));
      expect(prj.getSnapshot()).toMatchSnapshot();
      done();
    } catch (err) {
      done(err);
    }
  });


  it("simple_babel_prj(keep filename)", async done => {
    const rootDir = path.join(__dirname, "test-fixtures/simple_babel_prj");
    const prj = await createProject<TestProject>(TestProject, { rootDir });
    try {
      await rename(prj, path.join(rootDir, "src/core/target.js"), path.join(rootDir, "src/feat"));
      expect(prj.getSnapshot()).toMatchSnapshot();
      done();
    } catch (err) {
      done(err);
    }
  });

  it("babel_root_import_prj", async done => {
    const rootDir = path.join(__dirname, "test-fixtures/babel_root_import_prj");
    const prj = await createProject<TestProject>(TestProject, { rootDir });
    expect(prj.getFileMappingOptions().rootImport).toBeTruthy();
    try {
      await rename(prj, path.join(rootDir, "src/common/util.js"), path.join(rootDir, "src/feat-b/util.js"));
      expect(prj.getSnapshot()).toMatchSnapshot();
      done();
    } catch (err) {
      done(err);
    }
  });

  it("babel_root_import_prj_flatten", async done => {
    const rootDir = path.join(__dirname, "test-fixtures/babel_root_import_prj_flatten");
    const prj = await createProject<TestProject>(TestProject, { rootDir });
    expect(prj.getFileMappingOptions().rootImport).toBeTruthy();
    try {
      await rename(prj, path.join(rootDir, "src/common/util.js"), path.join(rootDir, "src/feat-b/util.js"));
      expect(prj.getSnapshot()).toMatchSnapshot();
      done();
    } catch (err) {
      done(err);
    }
  });

  it("css_modules_prj", async done => {
    const rootDir = path.join(__dirname, "test-fixtures/css_modules_prj");
    const prj = await createProject<TestProject>(TestProject, { rootDir });
    try {
      await rename(prj, "src/components/Hoge", "src/components/Fuga");
      expect(prj.getSnapshot()).toMatchSnapshot();
      done();
    } catch (err) {
      done(err);
    }
  });
});
