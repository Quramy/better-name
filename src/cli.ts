#!/usr/bin/env node

import { createDefaultProject } from "./project";
import { rename } from "./rename";

async function main() {
  if (process.argv.length < 4) return;
  const prj = await createDefaultProject({ rootDir: process.cwd() });
  await rename(prj, process.argv[2], process.argv[3]);
}

main();
