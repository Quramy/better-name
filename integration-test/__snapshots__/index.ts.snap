// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration test babel_root_import_prj 1`] = `
"
// id: src/feat-a/index.js

import { awesome } from '~/feat-b/util';

awesome();

// id: src/feat-b/util.js

export function awesome() {
  return 'awesome!';
}"
`;

exports[`integration test simple_babel_prj 1`] = `
"
// id: src/core/dependee.js

export function util() {
  console.log('this is util func');
}

// id: src/core/index.js

import { someFn } from '../feat/dest';

export default function main() {
  someFn();
}

// id: src/feat/dest.js

import { util } from '../core/dependee';

export function someFn() {
  util();
  return 'test';
}"
`;
